plan_delegator/src/plan_delegator.cpp

```diff
@@ -14,65 +14,169 @@
+    {
+        // currently it only checks if maneuver list is empty
+        return !maneuver_plan->maneuvers.empty();
+    }
+
+    bool PlanDelegator::IsTrajectoryValid(const cav_msgs::TrajectoryPlan& trajectory_plan) const noexcept
+    {
+        // currently it only checks if trajectory contains less than 2 points
+        return !(trajectory_plan.trajectory_points.size() < 2);
+    }
+
+    bool PlanDelegator::IsManeuverExpired(const cav_msgs::Maneuver& maneuver, ros::Time current_time) const
+    {
+        return GET_MANEUVER_PROPERTY(maneuver, end_time) <= current_time;
+    }
+
+    cav_srvs::PlanTrajectory PlanDelegator::ComposePlanTrajectoryRequest(const cav_msgs::TrajectoryPlan& latest_trajectory_plan) const
+    {
+        auto plan_req = cav_srvs::PlanTrajectory{};
+        plan_req.request.maneuver_plan = latest_maneuver_plan_;
+        // set current vehicle state if we have NOT planned any previous trajectories
+        if(latest_trajectory_plan.trajectory_points.empty())
+        {
+            plan_req.request.vehicle_state.longitudinal_vel = latest_twist_.twist.linear.x;
+            plan_req.request.vehicle_state.X_pos_global = latest_pose_.pose.position.x;
+            plan_req.request.vehicle_state.Y_pos_global = latest_pose_.pose.position.y;
+        }
+        // set vehicle state based on last two planned trajectory points
+        else
+        {
+            cav_msgs::TrajectoryPlanPoint last_point = latest_trajectory_plan.trajectory_points.back();
+            cav_msgs::TrajectoryPlanPoint second_last_point = *(latest_trajectory_plan.trajectory_points.rbegin() + 1);
+            plan_req.request.vehicle_state.X_pos_global = last_point.x;
+            plan_req.request.vehicle_state.Y_pos_global = last_point.y;
+            auto distance_diff = std::sqrt(std::pow(last_point.x - second_last_point.x, 2) + std::pow(last_point.y - second_last_point.y, 2));
+            auto time_diff = (last_point.target_time - second_last_point.target_time) * MILLISECOND_TO_SECOND;
+            // this assumes the vehicle does not have significant lateral velocity
+            plan_req.request.vehicle_state.longitudinal_vel = distance_diff / time_diff;
+        }
+        return plan_req;
+    }
 
-                cav_srvs::PlanTrajectory plan_req;
-                plan_req.request.maneuver_plan = *plan;
-                if (trajectory_planners[first_maneuver_planner].call(plan_req))
+    bool PlanDelegator::IsTrajectoryLongEnough(const cav_msgs::TrajectoryPlan& plan) const noexcept
+    {
+        return (plan.trajectory_points.back().target_time - plan.trajectory_points.begin()->target_time) * MILLISECOND_TO_SECOND >= max_trajectory_duration_;
+    }
+
+    cav_msgs::TrajectoryPlan PlanDelegator::PlanTrajectory()
+    {
+        cav_msgs::TrajectoryPlan latest_trajectory_plan;
+        // iterate through maneuver list to make service call
+        for(const auto& maneuver : latest_maneuver_plan_.maneuvers)
```

Comment:
I don't see any explicit multithreading so far, so I'm not super concerned, but if there's any that I've missed it would be important to ensure that there's a lock on the vector this for-each loop is iterating.

Only adding this since the `-pthread` compile option has be a bit nervous.

Reply:
Yup. It is a single thread node. I just removed the pthread option.